bloomfilter/src/Data/BloomFilter/BitVec64.hs

```diff
@@ -49,6 +54,13 @@ unsafeIndex (BV64 bv) i =
 unsafeTestBit :: Word64 -> Int -> Bool
 unsafeTestBit w k = w .&. (1 `unsafeShiftL` k) /= 0
 
+{-# INLINE prefetchIndex #-}
+prefetchIndex :: BitVec64 -> Int -> ST s ()
+prefetchIndex (BV64 (P.Vector (I# off#) _ (ByteArray ba#))) (I# i#) =
+    --TODO: explain the shiftR 3 here, vs the shiftR 6 above
+    ST (\s -> case prefetchByteArray0# ba# (off# +# uncheckedIShiftRL# i# 3#) s of
+                s' -> (# s', () #))
```

Comment:
Yes the whole BitVec64 code only works properly on 64bit systems (and the same assumption is elsewhere in the bloomfilter code).

Reply:
So i# is an offset in bits, so shiftR 3 to get bytes, and shiftR 3 again to get word64s. And `unsafeIndex` uses `P.unsafeIndex` which works with word64 offsets (and so internally shifts left by 3 again to get a byte offset). So that one effectively does a shiftR 6 and then a shiftL 3. Note that these do not cancel out to give shiftR 3 because that would not zero out the low three bits to give an 8byte aligned address (for the Word64).

For the prefetch we just need a shiftL 3 because we need a byte offset, and we _do not need the byte offset to be 8byte aligned_ (like the `unsafeIndex` does).

Reply:
Makes sense. And `prefetchByteArray0#` would prefetch a machine word?

Reply:
Added a comment.